### ISSUE TITLE
Cash On Delivery - Order status = Processing instead of On-hold

### ISSUE THREAD

**Issue by jayb75** [woothemes/woocommerce#10522]
## EXPLANATION OF THE ISSUE

When ordering by "Cash On Delivery" method, the order status is set to "Processing" despite it has not be paid yet. The status should be "On-hold" for a non paid order.

## STEPS TO REPRODUCE THE ISSUE

Allow cash on delivery method, make a payment by this method and see the order status

**Comment by mikejolley** [comment]
Changed back in https://github.com/woothemes/woocommerce/issues/3102

**Comment by mikejolley** [comment]
I does kind of make more sense to make this processing. It's not waiting on payment variation, its ready to ship and when shipped you collect funds. On-hold is not quite suitable - that infers that the order cannot be fulfilled yet.

**Comment by harkalygergo** [comment]
Shorter answer:

```
// change COD payment method order status from processing to on-hold
add_action('woocommerce_thankyou_cod', 'action_woocommerce_thankyou_cod', 10, 1);
function action_woocommerce_thankyou_cod($order_id)
{
	$order = wc_get_order($order_id);
	$order->update_status('on-hold');
}
```

**Comment by user354** [comment]
@harkalygergo this code changes order status from "processing" to "on hold" but originally it is processing so user could order a gift certificate, choose cash on delivery and receive the gift certificate even though he has not paid!

**Comment by user354** [comment]
@mikejolley it seems that cash on delivery should have custom order status that means order has been received but payment has not. What's wrong with "pending payment". "Processing" status is not suitable since wc understands processing to mean paid (wc docs also define processing as "payment received").

What does the code update in 29d3515 do?

Thanks :)

**Comment by kaydeeweb** [comment]
@harkalygergo's answer worked for me. Thanks very much.

**Comment by izzycart** [comment]
@jayb75's Answer Worked for me. You Saved me from frustration. Thanks so much

**Comment by lomars** [comment]
There is a [dedicated filter hook](https://docs.woocommerce.com/wc-apidocs/source-class-WC_Gateway_COD.html#) that will make that easy. 

So you can use instead this simple code snippet instead: 

```
add_filter( 'woocommerce_cod_process_payment_order_status', 'change_cod_payment_order_status', 10, 1 );
function change_cod_payment_order_status( $order_status, $order ) {
	return 'on-hold';
}
```
Tested and works… [See it on StackOverFlow](https://stackoverflow.com/questions/53280800/plugin-of-woocommerce-sets-order-status-to-processing-instead-of-on-hold/53291919#53291919)

**Comment by Rana858** [comment]
> Since March 2016 code update in [29d3515](https://github.com/woocommerce/woocommerce/commit/29d35151a19ccdc2457a80aff1e20189c0f40a9b), there is a [dedicated filter hook](https://docs.woocommerce.com/wc-apidocs/source-class-WC_Gateway_COD.html#) that can be used in this simple way:
> 
> ```
> add_filter( 'woocommerce_cod_process_payment_order_status', 'change_cod_payment_order_status' );
> function change_cod_payment_order_status( $order_status, $order ) {
> 	return 'on-hold';
> }
> ```
> Tested and works… [See it on StackOverFlow](https://stackoverflow.com/questions/53280800/plugin-of-woocommerce-sets-order-status-to-processing-instead-of-on-hold/53291919#53291919)

i want to add conditional by shipping and billing state, how?

**Comment by sgeithman** [comment]
@Rana858

> i want to add conditional by shipping and billing state, how?

```
add_filter( 'woocommerce_cod_process_payment_order_status', 'change_cod_payment_order_status' );
function change_cod_payment_order_status( $order_status, $order ) {
	
	$order = new WC_Order($order_id); // order object
	
	$address_billing = $order->get_address('billing'); // returns an array
	$address_shipping = $order->get_address('shipping'); // returns an array

	/* ----- `$address_billing` and `$address_shipping` 
	Array
	(
		[first_name] => First
		[last_name] => Name
		[company] => Company Name
		[address_1] => 123 123rd Street
		[address_2] => Suite #123
		[city] => Phoenix
		[state] => AZ
		[postcode] => 85027
		[country] => US
		[email] => email@someaddress.com
		[phone] => 555-555-1212
	)
	*/
	
	if($address_billing['state'] === 'AZ') {
		return 'on-hold';
	}

	if($address_shipping['state'] === 'HI') {
		return 'failed';
	}

	// fail safe - return expected status
	return $order_status;

}
```

There is also `'woocommerce_payment_complete_order_status'` hook that works just like `'woocommerce_cod_process_payment_order_status'` except on actual payment and not `COD`

**Comment by lomars** [comment]
@sgeithman There are some mistakes in your code:

- The `$order_id` variable is not defined
- The `$order` variable _(The WC_Order object)_ is already included
- Better use `get_billing_state()` and `get_shipping_state()` methods instead that get directly billing and shipping state.
- Only billing state is really useful with COD payments as there is no need of shipping address.

So the correct code should be something like:

```
add_filter( 'woocommerce_cod_process_payment_order_status', 'change_cod_payment_order_status' );
function change_cod_payment_order_status( $order_status, $order ) {
	
	$billing_state = $order->get_billing_state();
	// $shipping_state = $order->get_shipping_state();
	
	if( in_array( $billing_state, array('AL', 'AZ', 'CA') ) {
		$order_status = 'on-hold';
	} elseif( in_array( $billing_state, array('AR', 'CO', 'CT') ) {
		$order_status = 'custom'; // a custom status for example
	}
	
	return $order_status; // Default status is "processing"
}
```

It should better work.

**Comment by sgeithman** [comment]
@lomars For me `$order` is "integer" (maybe woocommerce version?). Maybe a check is need ...

```
if(!is_object($order)) {
    $order = new WC_Order($order);
}
```

```
add_filter( 'woocommerce_cod_process_payment_order_status', 'change_cod_payment_order_status' );
function change_cod_payment_order_status( $order_status, $order_id ) {
    $order = new WC_Order($order_id); // create order object - needed for me
    ....
```

**Comment by lomars** [comment]
The variable` $order` is already the `WC_Order` Object instance, simply [look at the source code where this filter hook is located](https://docs.woocommerce.com/wc-apidocs/source-class-WC_Gateway_COD.html#279). So it's not an integer, but an Object…

**Comment by almostsultry** [comment]
I want to take this in the opposite direction. I want to automatically change the status to "completed" when a COD transaction is checked out AND a user with a specific role is logged in.

This would allow me to use the default WooCommerce cart experience to handle cash transactions at farmer's markets without needing to go back and remember specific transactions from a hectic and overheated day. This means no reconciliation is necessary other than totaling the cash drawer and filtering the order records.

Can anyone help me adapt this thread to that goal?
